---
title: "Tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

This tutorial will guide you through the main functions of the `RightSeuratTools` package. The package is designed to provide complementary tools to the `Seurat` package, such as plotting, data manipulation, and more. The package is still under development, and new functions will be added in the future.

## Installation

You can install the `RightSeuratTools` package from GitHub using the `devtools` package. If you don't have the `devtools` package installed, you can install it using the following command:

`install.packages("devtools")`

Be sure to also install in a similar fashion all other packages needed for this tutorial. Once you have `devtools` installed, you can install the `RightSeuratTools` package using the following command:

`devtools::install_github("Alexis-Varin/RightSeuratTools")`

## Loading the package

```{r}
library(RightSeuratTools)
```

## Data loading and preprocessing

For this tutorial, we will use the `pbmc3k` dataset, which contains single-cell RNA-seq data from peripheral blood mononuclear cells (PBMCs).

We will also build on `Seurat`'s own tutorial for `pbmc3k` to load and preprocess the data, available for reference [here](https://satijalab.org/seurat/articles/pbmc3k_tutorial).

First, download the [raw data](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).

Then, extract the compressed files and put the filtered_gene_bc_matrices folder in your working directory, which you can determine using `getwd()`.

We can then proceed in loading the data and preprocessing it using the `Seurat` package.

```{r}
library(Seurat)

# Load the pbmc3k dataset
pbmc.data <- Read10X(data.dir = "filtered_gene_bc_matrices/hg19/")

# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc

# To be continued...
```
