<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started with RightOmicsTools • RightOmicsTools</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with RightOmicsTools">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">RightOmicsTools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/RightOmicsTools.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/Visualization.html">Advanced Data Visualization</a></li>
    <li><a class="dropdown-item" href="../articles/GSEA.html">Creating gene signatures from GSEA pathways</a></li>
    <li><a class="dropdown-item" href="../articles/tradeSeq.html">Analysis of DEG along pseudotime</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Alexis-Varin/RightOmicsTools/issues">Issues</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Alexis-Varin/RightOmicsTools/discussions">Feedbacks</a></li>
<li class="nav-item"><a class="nav-link" href="../authors.html">Citation</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://orcid.org/0009-0001-3474-7263">About</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Alexis-Varin/RightOmicsTools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting Started with RightOmicsTools</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Alexis-Varin/RightOmicsTools/blob/main/vignettes/RightOmicsTools.Rmd" class="external-link"><code>vignettes/RightOmicsTools.Rmd</code></a></small>
      <div class="d-none name"><code>RightOmicsTools.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tutorial will guide you through the main functions of the
<code>RightOmicsTools</code> package, designed to provide complementary
tools to the analysis of single-cell RNA-seq data, such as plotting,
data manipulation, and more. The package is still under development, and
new functions will be added in the future.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>If you are using Windows, you first need to make sure you have Rtools
installed, as some dependencies needed for <code>RightOmicsTools</code>
require compilation. You may download Rtools from <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">here</a>.</p>
<p>You may install the package from GitHub using the
<code>devtools</code> package, which, if you don’t have it, may be
installed using the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span></code></pre></div>
<p>Once installed, you may install <code>RightOmicsTools</code> using
the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Alexis-Varin/RightOmicsTools"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<p>For this tutorial, we will use the <code>pbmc3k</code> dataset, which
contains single-cell RNA-seq data from peripheral blood mononuclear
cells (PBMCs), and made available by <a href="https://www.10xgenomics.com/datasets/3-k-pbm-cs-from-a-healthy-donor-1-standard-1-1-0" class="external-link">10X
Genomics</a>.</p>
<p>We will also build on the <code>Seurat</code> package’s
<code>pbmc3k</code> tutorial to preprocess it, available for reference
<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" class="external-link">here</a>.</p>
<p>Let’s start by loading all the packages we will be using in this
vignette:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alexis-varin.github.io/RightOmicsTools/">RightOmicsTools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>Next, we will load the <code>pbmc3k</code> dataset using our first
function from <code>RightOmicsTools</code>, <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Right_Data.html">Right_Data</a>,
which handles all data preparation for us, and check its contents:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Checking available datasets</span></span>
<span><span class="fu"><a href="../reference/Right_Data.html">Right_Data</a></span><span class="op">(</span>list.datasets <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Loading the data</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Right_Data.html">Right_Data</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Checking the contents of the object</span></span>
<span><span class="va">pbmc3k</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Available datasets:</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">"pbmc3k"</span>   a Seurat v4 object of 2700 cells by 13714 genes</span></span></code></pre>
<pre><code><span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13714 features across 2700 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13714 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  1 dimensional reduction calculated: umap</span></span></code></pre>
<pre><code><span><span class="co">#&gt; [1] "orig.ident"         "nCount_RNA"         "nFeature_RNA"      </span></span>
<span><span class="co">#&gt; [4] "seurat_annotations" "treatment"</span></span></code></pre>
<p>Since the object seems to contain a ‘data’ layer, a UMAP reduction
and annotations, and we are unsure to what extent is has been
preprocessed and processed, we will also test another function from
<code>RightOmicsTools</code>, <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Right_DietSeurat.html">Right_DietSeurat</a>,
which is a reworked version of <a href="https://satijalab.org/seurat/reference/dietseurat" class="external-link">DietSeurat</a>,
and will remove all the layers and slots from the object, leaving only
the counts layer from the RNA assay. We will also remove all meta.data
columns except for the ‘orig.ident’ metadata to truly start anew. <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Right_DietSeurat.html">Right_DietSeurat</a>
is highly customizable, check the documentation for more
information:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reducing the object to the counts layer and the 'orig.ident' metadata</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Right_DietSeurat.html">Right_DietSeurat</a></span><span class="op">(</span><span class="va">pbmc3k</span>, idents <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Checking the contents of the object</span></span>
<span><span class="va">pbmc3k</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13714 features across 2700 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span></code></pre>
<pre><code><span><span class="co">#&gt; [1] "orig.ident"   "nCount_RNA"   "nFeature_RNA"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h2>
<div class="section level3">
<h3 id="quality-control-and-cell-filtering">Quality control and cell filtering<a class="anchor" aria-label="anchor" href="#quality-control-and-cell-filtering"></a>
</h3>
<p>We will now move on to preprocessing. We start by calculating the
percentage of mitochondrial genes and plot these results alongside the
number of genes (nFeature_RNA) and counts (nCount_RNA) per cell:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate percentage of mitochondrial genes</span></span>
<span><span class="va">pbmc3k</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot QC</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nCount_RNA</span>, y <span class="op">=</span> <span class="va">nFeature_RNA</span>, col <span class="op">=</span> <span class="va">percent.mt</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span>,<span class="st">"green3"</span>,<span class="st">"yellow3"</span>,<span class="st">"red3"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"pbmc3k QC metrics"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Log10(counts)"</span>, y <span class="op">=</span> <span class="st">"Log10(number of genes)"</span>, col <span class="op">=</span> <span class="st">"% mito"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/mito-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Based on the plot above, we will filter out cells with more than 10%
mitochondrial genes as well as cells with less than 400 or more than
2500 genes:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter cells</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc3k</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">400</span> <span class="op">&amp;</span></span>
<span>                              <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span></span>
<span>                              <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Checking the contents of the object</span></span>
<span><span class="va">pbmc3k</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13714 features across 2595 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>Now that we have filtered the cells, we will move on to the next
step, which is normalizing, scaling the data and identifying highly
variable genes. For convenience, we use the same parameters as in the
<code>Seurat</code> package tutorial:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalizing the data with default parameters</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find highly variable genes with default parameters</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scaling all genes, by default it only scales the variable features</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<p>Next, we will perform dimensionality reduction and unsupervised
clustering. We again use the same parameters as in the
<code>Seurat</code> package tutorial:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PCA</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Clustering</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc3k</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc3k</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc3k</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we will plot the UMAP, which concludes the preprocessing
steps:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/umap-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Interestingly, compared to the <code>Seurat</code> package tutorial,
we are missing the platelet cluster, which is likely due to the fact
that we have a more drastic cutoff for the number of genes per cell, at
400 versus 200. This is a good example of how the preprocessing steps
can influence the clustering results.</p>
</div>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<div class="section level3">
<h3 id="markers">Markers<a class="anchor" aria-label="anchor" href="#markers"></a>
</h3>
<p>A growing number of methods exist to label single-cell clusters, from
using reference datasets with <code>SingleR</code> package or <a href="https://azimuth.hubmapconsortium.org/" class="external-link">Azimuth</a>, to querying
ChatGPT with <code>GPTCelltype</code> package, via using canonical
markers from the scientific literature. The vast majority of these
methods need a set of genes as input, and determining the most
differentially expressed genes (DEG) in each cluster is a common first
step.</p>
<p>For this purpose, we are going to use a function from
<code>RightOmicsTools</code>, <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Find_Annotation_Markers.html">Find_Annotation_Markers</a>,
which is a wrapper function around <a href="https://satijalab.org/seurat/reference/findmarkers" class="external-link">FindMarkers</a>.
It is designed by default to conveniently output a character vector of
the top 5 markers, chosen based on the highest average log fold change
in genes expressed in at least 25% of cells in each cluster, and with
mitochondrial, ribosomal and non-coding genes excluded, in order to
maximize the chances of finding canonical markers in an unsupervised
manner. It also provides many more parameters which can be used to
tailor the function to your specific needs:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Top 5 markers for each cluster, we will name each marker with cluster identity</span></span>
<span><span class="va">annotation.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Find_Annotation_Markers.html">Find_Annotation_Markers</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>                                              name.features <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display markers for each cluster</span></span>
<span><span class="va">annotation.markers</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Finding markers for cluster 0 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 1 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 2 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 3 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 4 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 5 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 6 against all other clusters</span></span>
<span><span class="co">#&gt; Finding markers for cluster 7 against all other clusters</span></span></code></pre>
<pre><code><span><span class="co">#&gt;  [1] "CCR7"     "LEF1"     "MAL"      "PIK3IP1"  "TCF7"     "FOLR3"   </span></span>
<span><span class="co">#&gt;  [7] "S100A12"  "S100A8"   "S100A9"   "CD14"     "AQP3"     "CD40LG"  </span></span>
<span><span class="co">#&gt; [13] "CORO1B"   "CD2"      "TRAT1"    "VPREB3"   "CD79A"    "TCL1A"   </span></span>
<span><span class="co">#&gt; [19] "FCRLA"    "FCER2"    "GZMK"     "GZMH"     "CCL5"     "CD8A"    </span></span>
<span><span class="co">#&gt; [25] "KLRG1"    "CKB"      "CDKN1C"   "MS4A4A"   "HES4"     "BATF3"   </span></span>
<span><span class="co">#&gt; [31] "AKR1C3"   "GNLY"     "GZMB"     "SH2D1B"   "SPON2"    "SERPINF1"</span></span>
<span><span class="co">#&gt; [37] "FCER1A"   "CLIC2"    "ENHO"     "CLEC10A"</span></span></code></pre>
<p>These markers may then be directly used for plotting. Since the
<code>Seurat</code> package tutorial uses <a href="https://satijalab.org/seurat/reference/doheatmap" class="external-link">DoHeatmap</a>,
we will compare it to its equivalent in <code>RightOmicsTools</code>, <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Cell_Heatmap.html">Cell_Heatmap</a>,
which is reworked using the <code>ComplexHeatmap</code> package instead
of the <code>ggplot2</code> package:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Cell_Heatmap.html">Cell_Heatmap</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>             features <span class="op">=</span> <span class="va">annotation.markers</span>,</span>
<span>             cluster.features <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>             show.idents.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/heatmap-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Let’s compare it to <code>Seurat</code> package’s default heatmap
function:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html" class="external-link">DoHeatmap</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>          features <span class="op">=</span> <span class="va">annotation.markers</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/doheatmap-1.png" width="768" style="display: block; margin: auto;"></p>
<p>While both heatmaps look similar aside from colors, which is on
purpose with default parameters, <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Cell_Heatmap.html">Cell_Heatmap</a>
truly shines in complexity by offering more customization options, such
as the possibility to cluster features, apply k-means, or split
identities by another metadata…</p>
</div>
<div class="section level3">
<h3 id="cell-annotation">Cell annotation<a class="anchor" aria-label="anchor" href="#cell-annotation"></a>
</h3>
<p>We will now annotate each cluster with its corresponding cell type
and display the UMAP:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Annotate clusters</span></span>
<span><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>,</span>
<span>                     <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"NK"</span>, <span class="st">"DC"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">RenameIdents</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">new.cluster.ids</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add metadata to Seurat object</span></span>
<span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">named_clusters</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">active.ident</span></span>
<span></span>
<span><span class="co"># Plot UMAP with cell types</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/annotation-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="cell-proportion">Cell proportion<a class="anchor" aria-label="anchor" href="#cell-proportion"></a>
</h3>
<p>Following cell annotation, one often want to visualize the proportion
of each cell type. While using <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/table" class="external-link">table</a>
on the identities gives this information, it is often tedious to
organize these data for plotting; <code>RightOmicsTools</code>
introduces <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Barplot_Cell_Proportion.html">Barplot_Cell_Proportion</a>,
which automates cell proportion and conveniently displays a bar plot,
with the possibility to group and/or split based on other metadata, and
many other parameters:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Barplot_Cell_Proportion.html">Barplot_Cell_Proportion</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/proportion-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="other-visualizations">Other visualizations<a class="anchor" aria-label="anchor" href="#other-visualizations"></a>
</h3>
<p>We will conclude by showing the last two visualization functions; <a href="https://alexis-varin.github.io/RightOmicsTools/reference/DotPlot_Heatmap.html">DotPlot_Heatmap</a>,
which is a reworked version of <a href="https://satijalab.org/seurat/reference/dotplot" class="external-link">DotPlot</a>, also
built from the <code>ComplexHeatmap</code> package instead of the
<code>ggplot2</code> package, and <a href="https://alexis-varin.github.io/RightOmicsTools/reference/Grid_VlnPlot.html">Grid_VlnPlot</a>,
which is a stacked version of <a href="https://satijalab.org/seurat/reference/vlnplot" class="external-link">VlnPlot</a> in a
square grid, both also offering many more options than their
<code>Seurat</code> package’s counterparts:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We will disable some parameters to be as close as possible from Seurat's DotPlot</span></span>
<span><span class="co"># Due to the number of features, we will also lower dots size and flip the axis</span></span>
<span><span class="fu"><a href="../reference/DotPlot_Heatmap.html">DotPlot_Heatmap</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>                features <span class="op">=</span> <span class="va">annotation.markers</span>,</span>
<span>                dots.size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                cluster.features <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                cluster.idents <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                rotate.axis <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/dotplot-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Comparing it to <code>Seurat</code> package’s default dotplot
function:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Seurat's DotPlot function doesn't work well with named features, we will remove names</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">annotation.markers</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/dotplot_seurat_display-1.png" width="768" style="display: block; margin: auto;"></p>
<p>And finally, the grid violin plot:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Grid_VlnPlot.html">Grid_VlnPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>             features <span class="op">=</span> <span class="va">annotation.markers</span><span class="op">)</span></span></code></pre></div>
<p><img src="RightOmicsTools_files/figure-html/grid_vln-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="going-further">Going further<a class="anchor" aria-label="anchor" href="#going-further"></a>
</h2>
<div class="section level3">
<h3 id="advanced-data-visualization">Advanced data visualization<a class="anchor" aria-label="anchor" href="#advanced-data-visualization"></a>
</h3>
<p>While we have shown the default parameters of most of the functions
of <code>RightOmicsTools</code>, there are many more options available
in each function; to explore the use cases of most of these, and to see
how far they can be customized, check the <a href="https://alexis-varin.github.io/RightOmicsTools/articles/Visualization.html">advanced
data visualization vignette</a>.</p>
</div>
<div class="section level3">
<h3 id="gene-signatures-from-gsea">Gene signatures from GSEA<a class="anchor" aria-label="anchor" href="#gene-signatures-from-gsea"></a>
</h3>
<p>Another interesting feature of <code>RightOmicsTools</code> is the
possibility to extract genes from pathways in the Gene Set Enrichment
Analysis (GSEA) database to create and visualize signatures. Check the
<a href="https://alexis-varin.github.io/RightOmicsTools/articles/GSEA.html">gene
signatures from GSEA vignette</a> to learn more about the various usages
of this function.</p>
</div>
<div class="section level3">
<h3 id="differentially-expressed-genes-in-pseudotime-analysis">Differentially expressed genes in pseudotime analysis<a class="anchor" aria-label="anchor" href="#differentially-expressed-genes-in-pseudotime-analysis"></a>
</h3>
<p>Finally, <code>RightOmicsTools</code> provides helper and
visualization functions to greatly speed up and facilitate differential
gene expression in pseudotime analysis with the <code>tradeSeq</code>
package, with a focus on a highly customizable heatmap visualization
function. Head over to the <a href="https://alexis-varin.github.io/RightOmicsTools/articles/tradeSeq.html">DEG
along pseudotime analysis vignette</a> for an in-depth look at these
functions.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alexis Varin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
